Read frequency and modeling_realm from CMOR tables

bouweandela opened the pull request:
- Read frequency and modeling_realm from CMOR tables
- Also removes duplicate code from data finder

Closes #595

valeriupredoi commented:
whoaa! this looks so much cleaner @bouweandela - just need to fix the test, man, I like it lots compared to what we had before. Please don't tell me you made all these changes this morning :grin:

bouweandela commented:
Thanks. Please check that it still works as expected though, because I moved a lot of code around.

valeriupredoi commented:
@bouweandela have no doubts it will work, I'll fire up the test preproc recipe when you done pushing stuffs just in case

bouweandela commented:
Hopefully everything works now.

schlunma commented:
Hi Bouwe, I tested some recipes with your changes and had problems with the replacement of `[latestversion]` in the input directories:

https://github.com/ESMValGroup/ESMValTool/blob/ca88c26e5856d516c5c7038cdeecba8566ef3f97/esmvaltool/_data_finder.py#L164

This line always yields `False` since `part1` is only a relative path, so the replacement never takes place and the tool doesn't find any data. You have to include `rootpath` here.

valeriupredoi commented:
```_get_input_fx_dirnames()``` is struggling a lot since when working with vars that don't share the mips with sftlf: Amon or sftof: Omon etc, the ```table_entry``` will return None

bouweandela commented:
By the way, I noticed that the ensemble is not always `r0i0p0` for fx files (as currently configured in config-developer.yml), I found these files on Jasmin:
```bash
/badc/cmip5/data/cmip5/output1/MOHC/HadGEM2-ES/historical/fx/atmos/fx/r1i1p1/files/areacella_20110315/areacella_fx_HadGEM2-ES_historical_r1i1p1.nc
/badc/cmip5/data/cmip5/output1/MOHC/HadGEM2-ES/historical/fx/atmos/fx/r1i1p1/files/orog_20110315/orog_fx_HadGEM2-ES_historical_r1i1p1.nc
/badc/cmip5/data/cmip5/output1/MOHC/HadGEM2-ES/historical/fx/atmos/fx/r1i1p1/files/sftlf_20110315/sftlf_fx_HadGEM2-ES_historical_r1i1p1.nc
/badc/cmip5/data/cmip5/output1/MOHC/HadGEM2-ES/historical/fx/atmos/fx/r1i1p1/v20110315/areacella/areacella_fx_HadGEM2-ES_historical_r1i1p1.nc
/badc/cmip5/data/cmip5/output1/MOHC/HadGEM2-ES/historical/fx/atmos/fx/r1i1p1/v20110315/orog/orog_fx_HadGEM2-ES_historical_r1i1p1.nc
/badc/cmip5/data/cmip5/output1/MOHC/HadGEM2-ES/historical/fx/atmos/fx/r1i1p1/v20110315/sftlf/sftlf_fx_HadGEM2-ES_historical_r1i1p1.nc
```

bouweandela commented:
Ok, I've fixed the reported issues and added some unit tests. Can you please test again?

If anyone has access to one of the DKRZ/ETHZ/SMHI/BSC filesystems or CMIP6/OBS/ana4mips/obs4mips/CCMVal1/CCMVal2/EMAC datasets, can you add test cases to tests/integration/data_finder.yml?

mattiarighi commented:
I can take care of ETHZ and DKRZ, but I will need some time since I'm travelling until end of next week.

bouweandela commented:
@jvegasbsc @zklaus Could you also add test cases for your institutes?

mattiarighi commented:
Or maybe @bjoernbroetz can do this for DKRZ / ETHZ?

valeriupredoi commented:
I will test this on CEDA-JASMIN-BADC-NORAD-... right now

valeriupredoi commented:
works damn well with this rather bitchy recipe with data retrieval from BADC:
```
datasets:
  - {dataset: MPI-ESM-LR,  project: CMIP5, exp: historical,  ensemble: r1i1p1,  start_year: 2000,  end_year: 2002}
  #- {dataset: NorESM1-M,   project: CMIP5, exp: historical,  ensemble: r1i1p1,  start_year: 2001,  end_year: 2005}
  - {dataset: HadGEM2-ES,   project: CMIP5, exp: historical,  ensemble: r1i1p1,  start_year: 2000,  end_year: 2002}

preprocessors:
  pp_rad:
    mask_landsea:
      mask_out: land
    regrid:
      target_grid: 1x1
      scheme: linear

diagnostics:
  radiation_Amon_all_JRA-55:
    description: "CMIP5 vs JRA-55"
    variables:
      tos: # TOA LW up clear sky
        preprocessor: pp_rad
        field: T3M
        mip: Omon
        fx_files: [sftlf, sftof, areacello]
      fgco2: # TOA LW up clear sky
        preprocessor: pp_rad
        field: T3M
        mip: Omon
        fx_files: [sftlf]
      pr: # TOA LW up clear sky
        preprocessor: pp_rad
        field: T2Ms
        mip: Amon
    scripts: null
```
cool stuff @bouweandela :beer:

schlunma commented:
Tested with the new [Anav13 recipe](https://github.com/ESMValGroup/ESMValTool/blob/Anav13JClim_v2/esmvaltool/recipes/recipe_anav13jclim.yml) including a `ocnBgchem` variable (`fgco2`, which could not be read before), works perfectly!

mattiarighi commented:
Can you please solve the conflict and the codacy issues?
I will then run a couple of quick tests with DKRZ and ETHZ drs.

bouweandela commented:
@mattiarighi Instead of running some quick tests manually, can you please add your test cases to tests/integration/data_finder.yml? That way we automate tests, so next time someone updates this file there is no need for manual testing.

bouweandela commented:
The remaining Codacy issues are either TODO's that should be addressed in another pull request, or documentation issues/false positives in `tests` which are not worth the effort to fix.

mattiarighi commented:
I've added DKRZ and ETHZ. 
Is there a way to test this before committing? 

I'm not sure about the `fx` files: why for BADC there is only one fx variable (`sftof`). It that enough or should we have more?

bouweandela commented:
Great!
> Is there a way to test this before committing?

Yes, just run `python setup.py test` to run the tests on your own computer.

> I'm not sure about the fx files: why for BADC there is only one fx variable (sftof). It that enough or should we have more?

I think it is enough, because we already test that it can find many different types of fx files in the default case. However, more test cases is better (so long as the tests/integration/data_finder.yml file stays reasonably small and can still be read by a human without a crazy amount of scrolling), so feel free to add more.

mattiarighi commented:
Looks fine.
2 tests are failing, but they are not related to data_finder (`TestNaming.test_avoid_casing_collisions` and `TestNaming.test_no_namelist`).

bouweandela commented:
The tests are passing on CircleCI, so the failing tests you mention are probably caused by some unversioned files in your local copy of the repository.

bouweandela commented:
I have created issues for the problem with the hardcoded fx file ensemble member #622 and adding test cases #623.

schlunma commented:
Tested the [Anav13 recipe](https://github.com/ESMValGroup/ESMValTool/blob/Anav13JClim_v2/esmvaltool/recipes/recipe_anav13jclim.yml) again, unfortunately, it fails now with the following error:

```
Traceback (most recent call last):
  File "/pf/b/b309141/anaconda3/envs/esmvaltool/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "~/ESMValTool/esmvaltool/_task.py", line 552, in _run_task
    return task.run()
  File "~/ESMValTool/esmvaltool/_task.py", line 185, in run
    self.output_files = self._run(input_files)
  File "~/ESMValTool/esmvaltool/preprocessor/__init__.py", line 284, in _run
    input_files, self.settings, self.order, debug=self.debug)
  File "~/ESMValTool/esmvaltool/preprocessor/__init__.py", line 209, in preprocess_multi_model
    debug)
  File "~/ESMValTool/esmvaltool/preprocessor/__init__.py", line 246, in preprocess
    result.append(function(item, **args))
  File "~/ESMValTool/esmvaltool/cmor/fix.py", line 93, in fix_metadata
    checker(cube).check_metadata()
  File "~/ESMValTool/esmvaltool/cmor/check.py", line 107, in check_metadata
    self.report_errors()
  File "~/ESMValTool/esmvaltool/cmor/check.py", line 125, in report_errors
    raise CMORCheckError(msg)
esmvaltool.cmor.check.CMORCheckError: There were errors in variable nbp:
nbp: does not match coordinate rank
```

This is not an issue of a specific dataset but of the variable `nbp`. Since it worked before e6d4feebb58d23c21463769f0b22c37f3b6a3ef0 and the error occurs because the `coordinates` attributes of `self._cmor_var` is empty,

https://github.com/ESMValGroup/ESMValTool/blob/60a89f7828025c599615bcb5932b1917a40fb333/esmvaltool/cmor/check.py#L249

I think the problem is somewhere in e6d4feebb58d23c21463769f0b22c37f3b6a3ef0, which modified the reading of the `coordinates`.
